### ISSUE TITLE
BUG: Matrix multiplication in windows produces large numerical errors inconsistently in numpy >= 2

### ISSUE THREAD

**Issue by alasdairwilson** [numpy/numpy#27036]
### Describe the issue:

We discovered a simple test failing on windows only: https://github.com/sunpy/sunpy/issues/7754

I have investigated the symptoms of this in some detail but have not tried to find the cause: In short it seems like matrix multiplications with largeish numbers fails inconsistently in windows, and when it does fail it will fail a few times in a row before returning to normal. I would suspect openblas but the failure never occurs with numpy <2

### Reproduce the code example:

```python
import numpy as np
def test():
    # Test whether matrix multiplication involving a large matrix always gives the same answer
    # This indirectly tests whichever BLAS/LAPACK libraries that NumPy is linking to (if any)
    x = np.arange(500000, dtype=np.float64)
    src = np.vstack((x, -10*x)).T
    matrix = np.array([[0, 1], [1, 0]])

    expected = np.vstack((-10*x, x)).T  # src @ matrix
    
    mismatches = np.zeros(500, int)
    for i in range(len(mismatches)):
        result = src @ matrix
        mismatches[i] = (~np.isclose(result, expected)).sum()
        if mismatches[i] != 0:
            print(f"{mismatches[i]} mismatching elements in multiplication #{i}")
```

### Error message:

This bug is absolutely wild by the way:

With numpy 2.0.1:

```
16 mismatching elements in multiplication #22
316 mismatching elements in multiplication #33
28 mismatching elements in multiplication #53
307 mismatching elements in multiplication #100
32 mismatching elements in multiplication #201
32 mismatching elements in multiplication #267
288 mismatching elements in multiplication #272
1177 mismatching elements in multiplication #276
1596 mismatching elements in multiplication #289
32 mismatching elements in multiplication #298
16 mismatching elements in multiplication #314
3268 mismatching elements in multiplication #407
1403 mismatching elements in multiplication #419
64 mismatching elements in multiplication #446
```

with numpy 2.0.0
```
80 mismatching elements in multiplication #0
920 mismatching elements in multiplication #183
683 mismatching elements in multiplication #186
232 mismatching elements in multiplication #187
400 mismatching elements in multiplication #190
1217 mismatching elements in multiplication #195
1515 mismatching elements in multiplication #249
593 mismatching elements in multiplication #269
108 mismatching elements in multiplication #285
120 mismatching elements in multiplication #295
204 mismatching elements in multiplication #332
78 mismatching elements in multiplication #357
1816 mismatching elements in multiplication #380
164 mismatching elements in multiplication #396
48 mismatching elements in multiplication #492
```

with 1.26.4 it passes.

So it is a numpy 2 issue and not a 2.0.1 issue... it is also random:

Inserting a return and a break on the first time the matmul fails:

```
In [52]: test()
32 mismatching elements in multiplication #27

In [53]: test()
152 mismatching elements in multiplication #44

In [54]: test()
108 mismatching elements in multiplication #75

In [55]: test()
68 mismatching elements in multiplication #44

In [56]: test()
964 mismatching elements in multiplication #1

In [57]: test()
228 mismatching elements in multiplication #89

In [58]: test()
24 mismatching elements in multiplication #0
```

In a given set of multiplications the errors always have the same values.

```
In [92]: print([i for i in x if np.any(i != 0)])
[array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,   6948.]), array([-69480.,  
 6948.])]
```

```
In [94]: print([i for i in (res-ex) if np.any(i != 0)])
[array([-1289600.,   128960.]), array([-1289600.,   128960.]), array([-1289600.,   128960.]), array([-1289600.,   128960.]), array([-1289600.,   128960.]), array([-1289600.,   128960.]), array([-1289600.,   128960.]), array([-1289600.,   128960.])]
```

```
In [101]: print([i for i in (res-ex) if np.any(i != 0)])
[array([-208450.,   20845.]), array([-208450.,   20845.]), array([-208450.,   20845.]), array([-208450.,   20845.]), array([-208450.,   20845.]), array([-208450.,   20845.]), array([-208450.,   20845.]), array([-208450.,   20845.])]
```

As you can see they also have the weird symmetry where the first column is -10* the second column which I guess makes some twisted sense because thats what the calculation  is.

The errors are also always in the same region i.e. with indexs > 400k or so
![Figure_1](https://github.com/user-attachments/assets/2846c158-4a3a-47f0-8ff2-eaa9db1213f5)

![Figure_2](https://github.com/user-attachments/assets/6f4a109a-23ea-48a0-bb20-60dc960dc73f)

They also always fail multiple indexes in a row, e.g. you might have 20 or 50 or whatever failed calcualtions in a row but then it starts working again.

e.g.

```
[417292, 417293, 417294, 417295, 417296, 417297, 417298, 417299, 417300, 417301, 417302, 417303, 449292, 449293, 449294, 449295, 449296, 449297, 449298, 449299, 449300, 449301, 449302, 449303, 449400, 449401, 449402, 449403, 449404, 449405, 449406, 449407, 449408, 449409, 449410, 449411, 449412, 449413, 449414, 449415, 449416, 449417, 449418, 449419, 449420, 449421, 449422, 449423]
```
or 
```
[487736, 487737, 487738, 487739, 487740, 487741, 487742, 487743, 487760, 487761, 487762, 487763, 487764, 487765, 487766, 487767, 490490, 490491, 490492, 490493, 490494, 490495, 490496, 490497, 496838, 496839, 496840, 496841, 496842, 496843, 496844, 496845]
```

You can see in one of the plots there has been the same failure for 2 separate sequences of indices.

### Python and NumPy Versions:

Windows 10, cannot recreate this at all in linux.
```
python 3.11.8
numpy 2.0.0 or numpy 2.0.1 fail in the same manner 
```
This also failed on our CI on a windows-latest runner.
https://github.com/sunpy/sunpy/actions/runs/10072095800/job/27843572893#step:10:5160

### Runtime Environment:
```
[{'numpy_version': '2.0.1',
  'python': '3.11.8 (tags/v3.11.8:db85d51, Feb  6 2024, 22:03:32) [MSC v.1937 '
            '64 bit (AMD64)]',
  'uname': uname_result(system='Windows', node='DESKTOP-0EA2O83', release='10', version='10.0.19045', machine='AMD64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Haswell',
  'filepath': 'C:\\Users\\Alasdair\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\numpy.libs\\libscipy_openblas64_-fb1711452d4d8cee9f276fd1449ee5c7.dll',
  'internal_api': 'openblas',
  'num_threads': 8,
  'prefix': 'libscipy_openblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.27'}]
```
### Context for the issue:

Any matrix multiplication result could go wrong at any time in windows.

**Comment by mattip** [comment]
It does seem to be a problem with OpenBLAS and threading on windows. NumPy 1.26.4 uses `v0.3.23-293`. Numpy 2.0 and 2.0.1 use `v0.3.27.44`. Setting `OPENBLAS_NUM_THREADS=1` makes the test pass. I didn't see any relevant issues [searching for windows and threading](https://github.com/OpenMathLib/OpenBLAS/issues?q=is%3Aissue+windows+threading+is%3Aopen) on the OpenBLAS issue tracker, but maybe I need to search differently. @martin-frbg does this sound familiar?

**Comment by mattip** [comment]
I wonder how our CI passed :frowning_face:

**Comment by mattip** [comment]
I don't think this is related to the kernels chosen. The test still fails when I set `OPENBLAS_CORETYPE=PRESCOTT` (it was HASWELL).

**Comment by martin-frbg** [comment]
what does 0.3.27.44 correspond to, git-wise ? If post-0.3.27, any chance this could be related to the NAN handling changes in SCAL that I merged on June 29 ?

**Comment by ayshih** [comment]
For what it's worth, sunpy has this test because 4 years ago we encountered intermittent failures with OpenBLAS and multithreading on macOS (see https://github.com/sunpy/sunpy/issues/4290#issuecomment-676573472).  Those failures went away eventually with a subsequent release of OpenBLAS.

**Comment by martin-frbg** [comment]
Windows gets a lot less attention, and there _were_ a few PRs merged since 0.3.23 (whatever build your -293 is) and 0.3.27-whatever that (supposedly) improved multithreading on that platform. Notably https://github.com/OpenMathLib/OpenBLAS/pull/4359 - not trying to blame anyone, just that this would be the most serious change (merged in December,  so debuting in 0.3.26)

**Comment by martin-frbg** [comment]
@mseminatore

**Comment by mseminatore** [comment]
Is it possible to test this against an earlier (pre 0.3.26) OpenBLAS version to isolate whether this could be related to the Windows threading changes? @martin-frbg do we have existing validation tests that would/should show this issue?

Does this show up for any NTHREADS > 1?

Do we have a C repro for this yet? If not I will attempt to translate the Python to C and see if I can reproduce/debug it.

**Comment by martin-frbg** [comment]
I have just managed to set up a virtual environment for reproducing this (with a pip-installed numpy - I did not get conda to install anything other than mkl or netlib even if it claimed to have switched to openblas). It definitely happens with as few as 4 cores (though not on every run), but I have not yet gotten it to fail with just two. (Unfortunately I haven't figured out how to downgrade the openblas version, pip tells me there are no packages that match the version requirement)

**Comment by martin-frbg** [comment]
3 threads are definitely enough to reproduce it (on the fourth try), I still haven't seen a failure with 2 in about 20 runs so far

**Comment by mattip** [comment]
The openblas implementation is in <venv>/lib/site-packages/numpy.libs. You can try various [scipy-openblas64](https://pypi.org/project/scipy-openblas64/#history) versions. Note that pip-installing the package does not replace the DLL used by numpy: the package is only used when building numpy from source. You can copy the DLL from the site-package/scipy_openblas/lib directory, overwriting the one in numpy.libs (be sure to preserve the mangled name).

**Comment by mattip** [comment]
Note I am close to uploading a newer scipy-openblas64 based on latest OpenBLAS HEAD. You can download a pre-release from https://anaconda.org/scientific-python-nightly-wheels/scipy-openblas64/files. You want [scipy_openblas64-0.3.27.341.0-py3-none-win_amd64.whl](https://anaconda.org/scientific-python-nightly-wheels/scipy-openblas64/0.3.27.341.0/download/scipy_openblas64-0.3.27.341.0-py3-none-win_amd64.whl).

**Comment by martin-frbg** [comment]
@mattip thank you very much. No failures seen so far with 4 threads and the 0.3.24.95.1 version of the library copied over the original libscipy_openblas64_-fb1711452d4d8cee9f276fd1449ee5c7.dll in numpy.libs

**Comment by mattip** [comment]
@Siddharth-Latthe-07 is that a chatgpt-generated comment? Your comment does not add to the body of knowledge around the issue. We already know it is related to OpenBLAS, and that it is related to some thread-related problem inside OpenBLAS that only manifests itself on windows, so your analysis of the problem is not helpful. Converting the wheels we ship to another BLAS/LAPACK implementation is not something we are considering at this time.

**Comment by mseminatore** [comment]
I've seen this happen with dtype = float32 as well. @martin-frbg in validating that this doesn't happen with earlier versions have you by chance been able to grab a log file with server tracing enabled?

**Comment by mattip** [comment]
> a log file with server tracing enabled

How can I help provide a version with that enabled?

**Comment by mseminatore** [comment]
@mattip if OpenBLAS is built with SMP_DEBUG defined then the thread server will send debug logging to stderr. That may help hint at what is going on.

**Comment by martin-frbg** [comment]
#4577 was merged only after the 0.3.27 release so probably/hopefully not in that build yet. 
as I understood, the repro is from sunpy i.e. another level removed from numpy/scipy and was noticed there only because numpy/scipy switched from bundling OpenBLAS 0.3.23 to 0.3.27. the only other potential indication of trouble is a headsup from the conda packager that their build of 0.3.27 is surprisingly slow compared to earlier ones.

**Comment by martin-frbg** [comment]
(conda issue is https://github.com/conda-forge/openblas-feedstock/issues/160#issuecomment-2254470099)

**Comment by mseminatore** [comment]
> #4577 was merged only after the 0.3.27 release so probably/hopefully not in that build yet. as I understood, the repro is from sunpy i.e. another level removed from numpy/scipy and was noticed there only because numpy/scipy switched from bundling OpenBLAS 0.3.23 to 0.3.27. the only other potential indication of trouble is a headsup from the conda packager that their build of 0.3.27 is surprisingly slow compared to earlier ones.

Thank you @martin-frbg that is helpful to know! I will continue to review the code and attempt to debug the repro in pursuit of a working theory and a potential fix. Please let me know should you decide that you would prefer reverting the code given that this is not the first issue we've encountered and that this is having downstream impact. Regardless of whether we keep or revert the Windows code I hope we can expand test coverage to allow us to catch similar issues earlier in the future.

**Comment by martin-frbg** [comment]
I'll see if I can come up with anything tonight. We are already close to a month past my tentative release date for 0.3.28 due to various stability problems (including with my health), so I guess it would not be a problem to add another week in case it becomes obvious where this goes wrong.

**Comment by martin-frbg** [comment]
not sure if this output from a build with SMP_DEBUG is going to help (stopped the program after it reported the first couple of mismatches)
[debuglog.zip](https://github.com/user-attachments/files/16429045/debuglog.zip)

**Comment by mseminatore** [comment]
@mattip do you happen to know the magnitude of the errors? Is it just outside of epsilon or large? Trying to roughly characterize this as either lost work or error accumulation to guide the investigation.

**Comment by martin-frbg** [comment]
not sure if I read the initial post by @alasdairwilson correctly, but my impression from it is that the error is large and indicative of missed or misplaced work

**Comment by alasdairwilson** [comment]
> not sure if I read the initial post by @alasdairwilson correctly, but my impression from it is that the error is large and indicative of missed or misplaced work

Yeah @mseminatore the error is huge, when the matmul works the error is precisely 0 but when it doesn't work it can be 2e4, maybe more off (which is up to maybe 1e-1 rtol)

As for the repro, yeah because it was 2 packages downstream it wasn't being tested against nightly openblas builds or anything, it just failed when numpy upgraded their openblas. It doesn't have any dependency to reproduce except numpy itself.

**Comment by alasdairwilson** [comment]
I tried to recreate this in c but sadly failed. I am extremely incompetent at c, so it is likely I have just messed it up, in the hopes my effort wasn't entirely wasted:

```cpp
#include <stdio.h>
#include <stdlib.h>
#include <cblas.h>
#include <math.h>

void test() {
    int size = 1000000;
    int mismatches[500] = { 0 };

    double* x = (double*)malloc(size * sizeof(double));
    double* src = (double*)malloc(2 * size * sizeof(double));
    double* expected = (double*)malloc(2 * size * sizeof(double));
    double matrix[4] = { 0, 1, 1, 0 };
    double* result = (double*)malloc(2 * size * sizeof(double));

    for (int i = 0; i < size; i++) {
        x[i] = (double)i;
    }
    for (int i = 0; i < size; i++) {
        src[2 * i] = x[i];
        src[2 * i + 1] = -10 * x[i];
        expected[2 * i] = -10 * x[i];
        expected[2 * i + 1] = x[i];
    }

    for (int i = 0; i < 500; i++) {
        cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, size, 2, 2, 1.0, src, 2, matrix, 2, 0.0, result, 2);
        int mismatch_count = 0;
        for (int j = 0; j < 2 * size; j++) {
            if (fabs(result[j] - expected[j]) > 1e-9) {
                mismatch_count++;
            }
        }
        mismatches[i] = mismatch_count;
        if (mismatch_count != 0) {
            printf("%d mismatching elements in multiplication #%d\n", mismatch_count, i);
        }
    }

    free(x);
    free(src);
    free(expected);
}

int main() {
    int num_threads = openblas_get_num_threads();
    printf("Number of threads: %d\n", num_threads);
    test();
    return 0;
}
```

perhaps someone smarter than me can spot my mistake.

**Comment by mseminatore** [comment]
Here is the C code I am using. I've written it as a test case against my own private BLAS library. I have a private library so that I can iterate/test/evaluate code changes more efficiently.

So far, I am unable to reproduce this in my own BLAS library (which uses the same Windows threading code). That is not to say that there is not an issue with this code in OpenBLAS, only that it isn't an obvious and immediate failure elsewhere.

#include <stdio.h>
#include <string.h>
#include <time.h>
#include "test.h"
#include "cblas.h"

#define size 500000
#define checks 500
#define epsilon (1e-5 + 1e-8)

typedef float dtype;

dtype x[size];
dtype src[size][2];
dtype matrix[] = {0.0, 1.0, 1.0, 0.0};
dtype expected[size][2];
int mismatches[checks];
dtype result[size][2];

//------------------------------------------------------
//
//------------------------------------------------------
static int isclose()
{
	int count = 0;

	for (int i = 0; i < size; i++)
	{
		dtype delta = fabs(result[i][0] - expected[i][0]);
		if (delta > epsilon)
			count++;

		delta = fabs(result[i][1] - expected[i][1]);
		if (delta > epsilon)
			count++;
	}

	return count;
}

//
void clear_result()
{
	for (int i = 0; i < size; i++)
	{
		result[i][0] = 0.0;
		result[i][1] = 0.0;
	}
}

//------------------------------------------------------
//
//------------------------------------------------------
static void test_win_threads()
{
	for (int i = 0; i < size; i++)
	{
		x[i] = (dtype)i;
		src[i][0] = x[i];
		src[i][1] = -10.0 * x[i];
		expected[i][0] = -10.0 * x[i];
		expected[i][1] = x[i];
	}

	for (int i = 0; i < checks; i++)
	{
		clear_result();

		cblas_sgemm(CblasNoTrans, CblasNoTrans, size, 2, 2, 1.0, src, 2, matrix, 2, 1.0, result, 2);
		mismatches[i] = isclose();
		if (mismatches[i] > 0)
			printf("%d mismatching elements in multiplication %d", mismatches[i], i);
	}

	puts("done");
}

//------------------------------------------------------
//
//------------------------------------------------------
int main(int argc, char *argv[])
{
	test_win_threads();

	return 0;
}

**Comment by mattip** [comment]
Using the DLL from the 0.3.26.0.4 scipyopenblas64 wheels, this test and the one from https://github.com/numpy/numpy/issues/26643 both pass. That uses the OpenBLAS v0.3.26 tagged version.

**Comment by martin-frbg** [comment]
quoting myself from https://github.com/numpy/numpy/issues/27036#issuecomment-2250523317
> what does 0.3.27.44 correspond to, git-wise ? 

is this 0.3.27 "the release", or some git snapshot/cherrypicking ?

**Comment by mattip** [comment]
Ufortunately, that build also has a problem with windows threading, see https://github.com/numpy/numpy/issues/26240. That was the reason we updated to 0.3.27. I am not sure where we go from here.

**Comment by mattip** [comment]
> what does 0.3.27.44 correspond to, git-wise?

It should mean it is 44 commits past 0.3.27, but when looking at openblas-libs's commit `14ebf5d`, I see the pyproject.toml version is `0.3.27.44.1` but the OpenBLAS commit is exactly `v0.3.27`. That is a mistake and is misleading, sorry. 

The next jump in OpenBLAS version is in the 0.3.27.57 wheel, which uses OpenBLAS `de465ffd`.

**Comment by seberg** [comment]
Long shot, but @alasdairwilson one difference, is that NumPy would pass the first op as fortran order with `CblasTrans`.  Should confirm that `src = src.copy("C")` fixes it then in Python.

**Comment by mattip** [comment]
> Should confirm that `src = src.copy("C")` fixes it then in Python

No, I still am seeing problems even with this.

**Comment by mattip** [comment]
If I shrink the size of `x` to 300_000, I do not see the problem. Maybe a int32 overflow somewhere?

**Comment by martin-frbg** [comment]
> It should mean it is 44 commits past 0.3.27, but when looking at openblas-libs's commit `14ebf5d`, I see the pyproject.toml version is `0.3.27.44.1` but the OpenBLAS commit is exactly `v0.3.27`. That is a mistake and is misleading, sorry.
> 
> The next jump in OpenBLAS version is in the 0.3.27.57 wheel, which uses OpenBLAS `de465ffd`.

sorry I do not understand - does that mean your 0.3.27.44 happened to be OpenBLAS-0.3.27 exactly, or that its exact commit hash past 0.3.27 can no longer be determined ? (de465ffd would appear to be shortly _before_ 4577 landed so I think that other thread-handling change should not have any bearing if the sunpy folks used "0.3.27.44")

**Comment by martin-frbg** [comment]
> If I shrink the size of `x` to 300_000, I do not see the problem. Maybe a int32 overflow somewhere?

Good catch but that would be extra weird as mseminatore's work was entirely threading infrastructure not concerned with data sizes, and I do not recall any PR in the relevant timeframe that could have replaced a "blasint" with a standard int (rendering your regular INTERFACE64=1 build option useless). I had started bisecting, but ran into weird stability issues with the VM and/or compiler so need to redo everything...

**Comment by mattip** [comment]
> sorry I do not understand - does that mean your 0.3.27.44 happened to be OpenBLAS-0.3.27 exactly?

Yes

**Comment by mseminatore** [comment]
> > If I shrink the size of `x` to 300_000, I do not see the problem. Maybe a int32 overflow somewhere?
> 
> Good catch but that would be extra weird as mseminatore's work was entirely threading infrastructure not concerned with data sizes, and I do not recall any PR in the relevant timeframe that could have replaced a "blasint" with a standard int (rendering your regular INTERFACE64=1 build option useless). I had started bisecting, but ran into weird stability issues with the VM and/or compiler so need to redo everything...

The last issue was a bug in queue management where work submitted re-entrantly was lost. The way the join logic works `exec_blas_async_wait()` will hang if we don't process all the work that was submitted. So, I would expect lost work to result in a hang as before.

I am reviewing the log file that Martin captured to see if there are any hints there.

**Comment by mseminatore** [comment]
OK, looking over the log file, I am seeing the following pattern for work submission:

```
Server[ 0] Started.  Mode = 0x2003 M =   2 N=500000 K=  2
Server[ 1] Started.  Mode = 0x2003 M =   2 N=500000 K=  2
Server[ 2] Started.  Mode = 0x2003 M =   2 N=500000 K=  2
```
This suggests that each call to `cblas_dgemm()` which if I understand the repro code would be called from Python with M=2, N=500000 and K=2 was not decomposed into a set of smaller sub-tasks. Help me out @martin-frbg, is that expected behavior for a level3 GEMM call? 

I would expect the matrix to be converted into a number of sub-tiles/sub-tasks in a work_queue (such that each sub-tile A, B and C all fit in the L2$) but I am not familiar with the OpenBLAS L3 work model. My math says that a MxN matrix of order (2, 500000) of float64 would be 8MB. Therefore A, B, and C would be 8MB, 32b (2 * 2 * 8) and 8MB, thus 16MB in size, larger than any L2$ I've seen. But I am dealing with a family medical situation, it's late and I'm tired so I may be thinking wrongly about it.

**Comment by martin-frbg** [comment]
Saw this too, maybe it is the extreme imbalance of matrix dimensiions (and openblas failing to split on N). 
 I may have broken it myself in https://github.com/OpenMathLib/OpenBLAS/pull/4585 - an attempt to get better utilization of threads. If I got that wrong and it makes OpenBLAS go single-threaded, there would be nothing the level3 driver could do to split the workload. This should affect all platforms equally though, and would not explain why the problem only appears on Windows.
(It is very early morning here, and I have been dealing with an ongoing medical situation in my family too in recent months. I will try to look into this particular PR first thing in the morning. Eventually the bisect should find the cause, if it isn't this one either)

**Comment by martin-frbg** [comment]
sorry, so far I have only managed to trash my build environment - the ancient m2w64-gcc (5) from conda (that I am sure worked on another system on the weekend) suddenly throws internal compiler errors when building OpenBLAS. 
Trying to replace it with the conda-forge::gcc (14) ends in a mess of conflicting dependencies involving the msys2-conda-epoch package as well as "make", and a separate msys2/mingw64 build of OpenBLAS appears to be incompatible with the pip-installed numpy (although symbols look correct).

**Comment by mattip** [comment]
> Using the DLL from the 0.3.26.0.4 scipy-openblas64 wheels, this test and the one from https://github.com/numpy/numpy/issues/26643 both pass. That uses the OpenBLAS v0.3.26 tagged version.

Now I am getting myself confused. Rerunning the test with a 0.3.26 OpenBLAS fails, both when I use the DLL from the 0.3.26.4 wheel and also when I rebuild OpenBLAS v0.3.26. 

I am using this to rebuild OpenBLAS with the compilers from the rtools package. This installs gcc/gfortran 10.3

```
rem this needs rtools installed, run this in an admin cmd.exe
rem choco install -y rtools --no-progress --force --version=4.0.0.20220206
set BASH_PATH=c:\rtools40\usr\bin\bash.exe
set CHERE_INVOKING=yes
set INTERFACE64=1
set LDFLAGS=-lucrt -static -static-libgcc
set MSYSTEM=UCRT64
%BASH_PATH% -lc make BINARY=64 DYNAMIC_ARCH=0 USE_THREAD=1 USE_OPENMP=0 NUM_THREADS=24 NO_WARMUP=1 NO_AFFINITY=1 CONSISTENT_FPCSR=1 BUILD_LAPACK_DEPRECATED=1 TARGET=PRESCOTT BUFFERSIZE=20 'LDFLAGS=-lucrt -static -static-libgcc -Wl,--defsym,quadmath_snprintf=snprintf' 'COMMON_OPT=-O2 -march=x86-64 -mtune=generic -fno-asynchronous-unwind-tables' 'FCOMMON_OPT= -O2 -march=x86-64 -mtune=generic -fno-asynchronous-unwind-tables -frecursive -ffpe-summary=invalid,zero -fdefault-integer-8' MAX_STACK_ALLOC=2048 INTERFACE64=1 SYMBOLSUFFIX=64_ LIBNAMESUFFIX=64_ SYMBOLPREFIX=scipy_ LIBNAMEPREFIX=scipy_ FIXED_LIBNAME=1
```

Then I can copy the DLL over the one used by numpy.

Unfortunately, the prefixing does not work correctly pre-0.3.26, so I will continue bisecting backwards from 0.3.26 using NumPy 1.26.4 (which does not have any `scipy_` function and dll prefixes but does use the `64_` suffixes).

**Comment by mattip** [comment]
I have 0.3.26 building for NumPy 1.26.4 (using only `SYMBOLSUFFIX`) and it fails the test. Now bisecting further back...

**Comment by martin-frbg** [comment]
Interesting, that would mean I am in the clear with my GEMM thread count change from PR 4585. Thanks for the chocolatey pointer, looks like I'm up and running with that again now.

**Comment by mattip** [comment]
I have a good commit with v0.3.25. Now I can bisect. 
> Bisecting: 91 revisions left to test after this (roughly 7 steps)"

**Comment by mattip** [comment]
Hmm. good is at `d32f38fb`, bad is at `302ca7edc7`, PR https://github.com/OpenMathLib/OpenBLAS/pull/4359 is inside the bisect range. I won't be able to finish the bisect today (each step takes an hour), but that is the most probable candidate of the remaining revisions. Maybe worthwhile staring at that a bit.

**Comment by martin-frbg** [comment]
thanks - my bisect is four steps from completion, I am also building a current HEAD with 4359 manually reverted now "just in case".

**Comment by martin-frbg** [comment]
Indeed HEAD with the changes from 4359 reverted appears to work fine - whether that means the problem is *in* 4359 or only uncovered by it remains to be seen. (Bisect still needs two steps but appears to be gravitating towards the same result)

**Comment by mseminatore** [comment]
Thank you @mattip abd @martin-frbg for the bisect work. I am sorry that I can’t more actively participate (for personal reasons) but will continue inspecting the code to see if I can identify a cause and fix.

**Comment by martin-frbg** [comment]
I think it is equally possible that the actual problem lies somewhere in driver/level3/level3_thread.c (though it has its own CriticalSection), but I lack the Windows developer experience to assess this

**Comment by mseminatore** [comment]
I haven’t looked very closely at that code but I will take a look.

**Comment by mattip** [comment]
Any further thoughts about #https://github.com/OpenMathLib/OpenBLAS/pull/4835 and whether NumPy should build its OpenBLAS with that PR as a patch to move the NumPy2.1 release forward?

**Comment by martin-frbg** [comment]
not sure if that counts as thoughts, but my current intention is to merge 4835 in order to get the 0.3.28 release out in the next couple of days (probably not today though)

**Comment by mattip** [comment]
I backported the changes in OpenBLAS's 4835 in the 0.3.27.44.4 version of scipy-openblas, and used those wheels in #27140. I also added the test to prevent this from regressing again. Closing. Please reopen or open a new issue if I missed something.

**Comment by charris** [comment]
There was another failure of the test for this on Python 3.11: https://github.com/numpy/numpy/actions/runs/20146759701/job/57829132899
```
  E               AssertionError: unexpected result from matmul, probably due to OpenBLAS threading issues
  E               assert False
  
  expected   = array([[-0.00000e+00,  0.00000e+00],
         [-1.00000e+01,  1.00000e+00],
         [-2.00000e+01,  2.00000e+00],
         ...99997e+06,  4.99997e+05],
         [-4.99998e+06,  4.99998e+05],
         [-4.99999e+06,  4.99999e+05]], shape=(500000, 2))
  i          = 0
  matrix     = array([[0, 1],
         [1, 0]])
  mismatches = np.int64(13068)
```

**Comment by mseminatore** [comment]
I guess we can safely rule out [PR #4359](https://github.com/OpenMathLib/OpenBLAS/pull/4359) for this one.

**Comment by mattip** [comment]
Wierd. That seems to be the only failure since this issue was closed, right? And it does not fail consistently.

**Comment by mseminatore** [comment]
It's the only issue that I've seen since then. I felt this at the time when 4359 was in play, I think we need some stress tests to help find low probability and edge case usage issues. I think it was ~9 months after 4359 was merged before an issue was found. 

Since we never found the root cause I don't know that we know for certain whether 4359 was cause or whether the increased efficiency (removing millions of kernel object create/deletes) on Windows helped to raise the probability of a latent issue.

If there is interest in expanding stress test coverage, I may be interested in helping. I would still like to see the Windows server code improved and believe that is necessary before proposing fundamental changes again.

**Comment by mattip** [comment]
> If there is interest in expanding stress test coverage

We could add more tests to OpenBLAS itself or to MacPython/openblas-libs (the build of OpenBLAS used by SciPy and NumPy). FWIW, I added the test at the top of this issue to NumPy in #27140. What do you suggest for testing?
